apis/pipelines/hub/pipeline_webhook.go

```diff
@@ -1,11 +1,137 @@
 package v1beta1
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/samber/lo"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
-func (p *Pipeline) SetupWebhookWithManager(mgr ctrl.Manager) error {
+func NewPipelineValidatorWebhook(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(p).
+		For(&Pipeline{}).
+		WithValidator(
+			&PipelineValidator{
+				reader: mgr.GetClient(),
+			},
+		).
 		Complete()
 }
+
+//+kubebuilder:webhook:path=/validate-pipelines-kubeflow-org-v1beta1-pipeline,mutating=false,failurePolicy=fail,sideEffects=None,groups=pipelines.kubeflow.org,resources=pipelines,verbs=create;update,versions=v1beta1,name=vpipeline.kb.io,admissionReviewVersions=v1
+//+kubebuilder:object:generate=false
+
+type PipelineValidator struct {
+	reader client.Reader
+}
+
+var _ webhook.CustomValidator = &PipelineValidator{}
+
+func (p *PipelineValidator) ValidateCreate(
+	ctx context.Context,
+	obj runtime.Object,
+) (admission.Warnings, error) {
+	return p.validate(ctx, obj)
+}
+
+func (p *PipelineValidator) ValidateUpdate(
+	ctx context.Context,
+	_, newObj runtime.Object,
+) (admission.Warnings, error) {
+	return p.validate(ctx, newObj)
+}
+
+func (p *PipelineValidator) ValidateDelete(
+	_ context.Context,
+	_ runtime.Object,
+) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (p *PipelineValidator) validate(
+	ctx context.Context,
+	obj runtime.Object,
+) (admission.Warnings, error) {
+	pipeline, ok := obj.(*Pipeline)
+
+	if !ok {
```

Comment:
It would be good if these could be separated into named funcs that describe what they are checking